### ISSUE TITLE
Concurrent connections

### ISSUE THREAD

**Issue by seansan** [mariusv/nginx-badbot-blocker#16]
Hi, is there a way we could "save" our judgement for a given IP? 

What I mean is that one says "If is Evil in nginx" and these if blocks are executed on every every connection

https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/

Suppose we  validate a referer or IP **per** connection made, meaning
- do the validation on the first connect
- after that allow the client to continue and dont validate on every request _again_

can this be done?

**Comment by mitchellkrogza** [comment]
The IF is evil is a bit misunderstood from what I understand and have read extensively, it seems that IF statements within location blocks are where most people make horrible mistakes. At a server context level they certainly do not give me any problems whatsoever. 

I have taken this current black listing method for nginx to a whole new level and will release it on github soon. I have it running on 2 production servers without a single glitch and it does not slow down nginx whatsoever simply because it is a centralised script loaded in the http context of the main nginx conf so it is in memory and immediate … always. I also update it daily with new bad referers and it has grown quite extensively from this current blocker, the entire script is a mere 52kb so it really is lightweight as ever. 

With a carefully constructed set of rules there is no false positives or any good IP’s blocked out by mistake. The key is testing, testing and testing some more and I have spent well over 280 hours now building a new nginx blocker which has been tested thoroughly and I watch logs daily to make 100% sure. I even picked up some false positives in this current version and also some user agent which should not be blocked at all because they are used by certain very good bots.

I have also built in rate limiting and anti DDOS protection and white lists for known good IP’s like my own and all of Google’s, Bing and Yahoo’s IP ranges. 

I personally don’t think Nginx even does a re-validate each time, it validates once and then if that IP breaks some other rule like it starts flooding you with 500 requests per seconds then it gets rate limited. With my blocker running or not running there is absolutely no speed difference on loading of pages whatsoever, tested this extensively too.

KR
Mitchell



On 12 August 2016 at 1:43:10 PM, seansan (notifications@github.com) wrote:

Hi, is there a way we could "save" our judgement for a given IP?

What I mean is that one says "If is Evil in nginx" and these if blocks are executed on every every connection

https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/

Suppose we validate a referer or IP per connection made, meaning

do the validation on the first connect
after that allow the client to continue and dont validate on every request again
can this be done?
